### PR TITLE
Add Settle command to log transfers made in an Activity

### PR DESCRIPTION
**Pull request by podocarp**
## Settle debts
Settle debts with `settle p/1 p/2 e/[amount]` which means 1 pays 2 $amount. If $amount is missing, then the entire debt is settled. Bounds checking performed for 
- overpaying (reject command)
- wrong sequence (so 1 owes 2 but then command asks for 2 to pay 1 some more, reject)
- same person (1 paying 1, reject)

Minor changes to other things:
- Amount treats $0 as valid now. Checking for trivial expenses is done by the expense command instead;
- Activity class exposes a `getDebt(id, id)` function instead of playing with the transfermatrix and getting the ID and whatnot, might be useful for UI @Aulud;
- Expense serializes `isDeleted` field as well -- this was missed out previously;
- Abstracted some things around in `ExpenseCommand` so `SettleCommand` can ride on them;
- Settles integrate into algo;
- `getNonSettlementExpenses()`, `getSettlementExpenses()`, `getSolution()` bindings;
- Added new `Triplet` class to return the algo output to UI more easierly.

## Todo
- [x] Change UG
- [x] Add tests for `command` (algo done)
- [ ] Add tests for `parser` (algo done)

Also I noticed that in `Activity.java` we have `getParticipantsIds` and `getParticipantIds`. One returns `List<Int>` and the other `ArrayList<Int>`. Should clean that up.

### CODE REVIEW

File: src/main/java/seedu/address/model/activity/Expense.java [hunk 1]
```diff
@@ -10,8 +10,9 @@
 public class Expense {
     private final int personId;
     private int[] involvedIds;
-    private final Amount amount;
+    private Amount amount;
     private final String description;
+    private boolean isSettlement;
     private boolean isDeleted;
```
Comment (liakify) [inline src/main/java/seedu/address/model/activity/Expense.java]: Need to update javadoc comments

Also not sure if I am missing something in the rest of the PR but why is `amount` mutable now?

File: src/main/java/seedu/address/logic/commands/SettleCommand.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            Expense e = new Expense(payingId, amount, description, receivingId);
+            e.setIsSettlement(true);
+            activity.addExpense(e);
+        } catch (PersonNotInActivityException e) {
+            throw new CommandException(MESSAGE_MISSING_PERSON_DESCRIPTION);
+        }
+
+        String successMessage = String.format(MESSAGE_SUCCESS, payingName, receivingName, amount.value);
+
+        return new CommandResult(successMessage
```
Comment (liakify) [inline src/main/java/seedu/address/logic/commands/SettleCommand.java]: Need to update context so that screen refreshes

File: src/main/java/seedu/address/logic/commands/ExpenseCommand.java [hunk 3]
```diff
@@ -106,47 +150,14 @@ public CommandResult execute(Model model) throws CommandException {
             }
         }
 
-        if (persons.size() > 1) {
-            List<Person> personList = new ArrayList<>();
-            personList.add(payingPerson);
```
Comment (liakify) [inline src/main/java/seedu/address/logic/commands/ExpenseCommand.java:111]: Any particular reason for not including the `payingPerson` in the `personList`? Previously there were warnings if the paying person is duplicated in the list of other involved people, which I thought was good to have just so that users are aware.

File: src/main/java/seedu/address/model/activity/Activity.java [hunk 4]
```diff
@@ -295,10 +303,22 @@ public void addExpense(Expense expense) throws PersonNotInActivityException {
...
         // We update the balance sheet
-        double splitAmount = amount / (involved.length + 1);
+        double splitAmount;
+        if (expense.isSettlement()) {
+            double debt = getOwed(involved[0], payer);
+            if (debt < 0) {
+                return;
+            } else if (amount > debt || amount == 0) {
+                splitAmount = debt;
+                System.out.println("ASD");
```
Comment (liakify) [inline src/main/java/seedu/address/model/activity/Activity.java]: Should remove before merging

File: docs/UserGuide.adoc [hunk 5]
```diff
@@ -198,6 +198,44 @@ e.g. `expense p/John Doe e/15 p/John Doe d/Cab` +
...
+Settles the debt between two people within the currently viewed activity.
+Specify two people, and optionally an amount `AMOUNT_PAID` to settle.
+There are two options available:
+* `AMOUNT_PAID` is not provided: then the command will automatically calculate the minumum needed to settle the debt and use that as the amount.
+* `AMOUNT_PAID` is provided: the command will use this amount.
+
+The first person specified `PERSON1` will pay `PERSON2`.
+
+`PERSON1` and `PERSON2` can be any part of the names of the activity participants you wish to target.
+Generally it is similar to how you would add an ordinary expense.
```
Comment (liakify) [inline docs/UserGuide.adoc:213]: It is actually the same cause you reused expense's person searching method

File: src/main/java/seedu/address/logic/parser/ExpenseCommandParser.java [hunk 6]
```diff
@@ -37,8 +37,12 @@ public ExpenseCommand parse(String args) throws ParseException {
         if (amounts.size() != 1) {
             throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, ExpenseCommand.MESSAGE_USAGE));
         }
+        Amount amount = amounts.get(0);
+        if (amount.value == 0) {
```
Comment (liakify) [inline src/main/java/seedu/address/logic/parser/ExpenseCommandParser.java]: Should use `isValidAmount` instead to validate, currently `expense p/p1 e/0.001` passes.

Also should update `ExpenseCommandParserTest.java` to include the edge cases that were previously covered by `AmountTest.java`, i.e. `assertFalse(Amount.isValidAmount(0)); // Not positive` and `assertFalse(Amount.isValidAmount(0.004999)); // Rounds to 0`.

File: src/test/java/seedu/address/model/activity/AmountTest.java [hunk 7]
```diff
@@ -16,9 +16,7 @@ public void constructor_invalidAmount_throwsIllegalArgumentException() {
     @Test
     public void isValidAmount() {
         // invalid amounts
-        assertFalse(Amount.isValidAmount(0)); // Not positive
         assertFalse(Amount.isValidAmount(-1)); // Negative int
-        assertFalse(Amount.isValidAmount(0.004999)); // Rounds to 0
```
Comment (liakify) [inline src/test/java/seedu/address/model/activity/AmountTest.java]: Refer to comment for `ExpenseCommandParser.java`

**Review by liakify — CHANGES_REQUESTED** [review]
1. Haven't thought through closely what's the best way to do this, but it would be good to be able to have settlement "expenses" made visually distinct from normal expenses cause its possible for users to type in the exact same description as what a settlement expense has, and it can potentially be confusing.

`settle p/p1 p/p2 e/10` is currently indistinguishable from `expense p/p2 e/20 p/p1 d/p1 paid p2`

2. Right now it seems like you're using `Amount(0)` as a way of flagging to `SettleCommand` to simply pay off all the debt. Rightfully I would think that there should be a separate class from `Expense`, e.g. `Payment` or something, and then have both `Expense` and `Payment` extend/implement `Transaction`. 

But non-essential cause we're probably not touching this code once this module is over.

File: src/main/java/seedu/address/logic/commands/SettleCommand.java (hunk 2 above)
Comment (JohnNzj) [reply src/main/java/seedu/address/logic/commands/SettleCommand.java]: Yes like what ^ said, it does not refreshes the screen, have to use the other commandresult constructor i guess

**Comment by Aulud** [comment]
>     1. Haven't thought through closely what's the best way to do this, but it would be good to be able to have settlement "expenses" made visually distinct from normal expenses cause its possible for users to type in the exact same description as what a settlement expense has, and it can potentially be confusing.
> 
> 
> `settle p/p1 p/p2 e/10` is currently indistinguishable from `expense p/p2 e/20 p/p1 d/p1 paid p2`
> 
>     1. Right now it seems like you're using `Amount(0)` as a way of flagging to `SettleCommand` to simply pay off all the debt. Rightfully I would think that there should be a separate class from `Expense`, e.g. `Payment` or something, and then have both `Expense` and `Payment` extend/implement `Transaction`.
> 
> 
> But non-essential cause we're probably not touching this code once this module is over.

We should try to implement more specific classes (e.g. use the `Amount` class for individual expense items and a `Transfer` class for required and actual transfers made) to distinguish between expenses and transfers.

Preferably, the `Activity` class should expose methods such as `getRequiredTransfers` so that it can retrieve a `List<Transfer>` for transfers yet to be made (we can lazily calculate this), and a `getCompletedTransfers` that will return a `List<Transfer>` for transfers that were actually made.

Otherwise, it is difficult for the UI component to be able to distinguish between expenses and actual transfers if they are all grouped together and treated as an actual `Expense`.

I could move some code from the UI component back into `Activity`, since right now it is doing some of the computational work.

File: src/main/java/seedu/address/model/activity/Activity.java (hunk 4 above)
Comment (JohnNzj) [reply src/main/java/seedu/address/model/activity/Activity.java]: Remember to remove the println

**Review by JohnNzj — COMMENTED** [review]
Mostly fine with only a few minor things to keep in mind, refer to comments

**Comment by liakify** [comment]
Currently settlement is reusing `Amount` and `Expense` to pay off debts. However, due to constraints on `Amount`, such bugs can occur where users are unable to directly settle amounts that exceed 1000000.

The program simply doesn't do anything which may confuse the user, and you can only do `settle p/tom p/you e/1000000` to pay off at most $1m at a time.
image

**Comment by podocarp** [comment]
> Currently settlement is reusing `Amount` and `Expense` to pay off debts. However, due to constraints on `Amount`, such bugs can occur where users are unable to directly settle amounts that exceed 1000000.

What do you suggest we do? To be honest I don't get the upper bound on `Amount`. If `Amount` has an upper bound, we should cap the total expenses in the whole activity, not just one expense in that activity. I for one am for removing the upper bound altogether.

File: src/main/java/seedu/address/logic/commands/ExpenseCommand.java (hunk 3 above)
Comment (podocarp) [reply src/main/java/seedu/address/logic/commands/ExpenseCommand.java:111]: It's checked for in line 109

File: src/main/java/seedu/address/model/activity/Activity.java (hunk 4 above)
Comment (podocarp) [reply src/main/java/seedu/address/model/activity/Activity.java]: Wow

File: src/main/java/seedu/address/logic/parser/ExpenseCommandParser.java (hunk 6 above)
Comment (podocarp) [reply src/main/java/seedu/address/logic/parser/ExpenseCommandParser.java]: ParserUtil catches those already.

File: src/test/java/seedu/address/model/activity/AmountTest.java (hunk 7 above)
Comment (podocarp) [reply src/test/java/seedu/address/model/activity/AmountTest.java]: Not sure what you mean.

**Comment by podocarp** [comment]
> Preferably, the `Activity` class should expose methods such as `getRequiredTransfers` so that it can retrieve a `List<Transfer>` for transfers yet to be made (we can lazily calculate this), and a `getCompletedTransfers` that will return a `List<Transfer>` for transfers that were actually made.

Done. See `getNonSettlementExpenses()`, `getSettlementExpenses()`, `getSolution()`

File: src/main/java/seedu/address/model/activity/Expense.java (hunk 1 above)
Comment (podocarp) [reply src/main/java/seedu/address/model/activity/Expense.java]: It was some old hack, reverted.

File: src/main/java/seedu/address/logic/commands/ExpenseCommand.java (hunk 3 above)
Comment (liakify) [reply src/main/java/seedu/address/logic/commands/ExpenseCommand.java:111]: But currently if you do `expense p/p1 e/10 p/p1` there is actually no warning shown cause the warnings are determined by this code segment and `personList` now doesn't contain the paying person ID
```
if (personList.contains(personPriKey)) {
    warningMessage.append(String.format(WARNING_DUPLICATE_PERSON, person.getName()));
    continue;
}
```

File: src/main/java/seedu/address/model/activity/Amount.java [hunk 8]
```diff
@@ -27,7 +28,7 @@ public Amount(double amount) {
      */
     public static boolean isValidAmount(double test) {
         double d = Math.round(test * 100) / 100.0;
-        return d > 0 && d <= 1000000;
+        return d >= 0 && d <= 1000000;
     }
```
Comment (liakify) [inline src/main/java/seedu/address/model/activity/Amount.java]: I think we should do this instead to clean up all the inconsistencies in acceptable amounts properly:
```
public Amount(double amount) {
    checkArgument(isValidAmount(amount), MESSAGE_POSITIVE_ONLY);
    value = amount;
}
public static boolean isValidAmount(double test) {
    return test >= 0;
}
public static boolean isValidExpenseAmount(double test) {
    double d = Math.round(test * 100) / 100.0;
    return d > 0 && d <= 1000000;
}
```


1. `isValidAmount` would represent acceptable object values and become in line with the corresponding `isValidXXX` methods in all the other field classes (`Name`, `Phone` etc.).
Constraint: value >= 0

2. `isValidExpenseAmount` would be used at the parser stage to process valid expense amounts. 
Constraint: 0.01 <= value after rounding to 2dp <= 1000000

3. Might want to have a `isValidSettlementAmount` that has slightly different conditions used for settlement parsing if you want different constraints on acceptable settlement amounts
Constraint: value > 0 (upper bound is covered by the fact that user cannot settle more than existing debt)

Admittedly not the cleanest OOP wise, but I think it's decent enough for the final deadline since we are unlikely to have any more new 'types' of expenses/amounts. 

On a more serious note is there any way to avoid using `Amount(0)` so we don't need to purposely split valid settlement and normal valid amounts?

**Comment by Aulud** [comment]
> > Preferably, the `Activity` class should expose methods such as `getRequiredTransfers` so that it can retrieve a `List<Transfer>` for transfers yet to be made (we can lazily calculate this), and a `getCompletedTransfers` that will return a `List<Transfer>` for transfers that were actually made.
> 
> Done. See `getNonSettlementExpenses()`, `getSettlementExpenses()`, `getSolution()`

Nice!! Thanks haha

File: src/main/java/seedu/address/logic/parser/ExpenseCommandParser.java (hunk 6 above)
Comment (liakify) [reply src/main/java/seedu/address/logic/parser/ExpenseCommandParser.java]: `ParserUtil` uses `isValidAmount` which allows amounts rounded to 0 to pass through

**Review by liakify — APPROVED** [review]
